src/libraries/System.Net.Http/src/System/Net/Http/HttpMessageInvoker.cs

```diff
@@ -14,6 +17,26 @@ public class HttpMessageInvoker : IDisposable
         private volatile bool _disposed;
         private readonly bool _disposeHandler;
         private readonly HttpMessageHandler _handler;
+        private Meter? _meter;
+        private HttpMetrics? _metrics;
+
+#pragma warning disable CS3003 // Type is not CLS-compliant
+        public Meter Meter
+        {
+            // TODO: Should the Meter and HttpMetrics be static and shared by default?
+            get => _meter ??= new Meter("System.Net.Http");
+            set
```

Comment:
I was expecting IMeterFactory to be in Microsoft.Extensions.Metrics. That's part of why I chose this approach.

The other reason for exposing the meter is it makes consuming the metrics for just one client quite simple:

```csharp
var invoker = new HttpMessageInvoker();
var currentRequestsRecorder = new InstrumentRecorder<double>(invoker.Meter, "request-duration");
```

But that only works if each client/invoker has its own Meter and counters copy, which might not be desirable.

Reply:
> I was expecting IMeterFactory to be in Microsoft.Extensions.Metrics and so not available here

Initially I imagined it there too. Do you feel like it is important for IMeterFactory to not be in System.Diagostics.Metrics instead?

> The other reason for exposing the meter is it makes consuming the metrics for just one client quite simple:

If users were injecting a factory instead the pattern wouldn't be much different:

```C#
var IMeterFactory factory = ...
var invoker = new HttpMessageInvoker();
invoker.Factory = factory; // separately, maybe we should be using a constructor overload rather than a property setter?
var currentRequestsRecorder = new InstrumentRecorder<double>(factory, "System.Net.Http", "request-duration");
```

If the app developer wasn't injecting the factory and using the default Meter then I wouldn't expect them to write tests for that as it would effectively just be testing runtime behavior and not their own code.

Reply:
It can also be a `Func<Meter>` on `HttpMessageInvoker`, so then we don't have to worry about the dependencies, and `IMeterFactory` can live in `Microsoft.Extensions.*` where the DI-friendly things go usually.

Reply:
Logically `IMeterFactory` naturally belongs to the Microsoft.extensions more because we'll have another extension method to register this interface to the DI. But technically nothing can stop us having it in DianosticSource library. I am thinking in the future when exposing more stuff like MeterBuilder or anything else, at that time it will look strange having IMeterFactory in one place and other related interfaces in other place.

Reply:
> It can also be a Func\<Meter\> on HttpMessageInvoker

`Func<string,Meter>` is the signature we'd need for a similar effect as IMeterFactory, but thats no obstacle. My other motivation for IMeterFactory instead of something like Func is so that .NET types are following the same guidance we are planning to give to everyone else. Its not written up in our official docs yet, but the basic pattern is:

1. Define a parameter on your constructor of type IMeterFactory
2. Use the factory to create the Meter.

While other patterns are certainly possible (such as passing in a Meter or passing in a Func<string,Meter>), many of them aren't as easy to use with DI. I don't expect folks will be creating HttpMessageInvokers from a DI container so its fair to argue it doesn't matter if we pick a pattern that is DI-friendly. I think the counter-argument is that if the DI-friendly pattern becomes a convention then devs will anticipate that pattern regardless whether the type is created from a DI-container and APIs that don't follow the pattern feel unexpected and awkward. For example most people expect that if your type logs via ILogger then your constructor has an ILogger<T> constructor parameter. Writing a class that has a method `SetLogger(ILogger)` or a constructor parameter `Func<ILogger>` functionally would work too, but it looks unusual because it doesn't follow the convention.

> Logically IMeterFactory naturally belongs to the Microsoft.extensions

Tarek and I chatted earlier. I agree that it felt intuitive for me as well initially for IMeterFactory to be in the Microsoft.Extensions namespace, but I don't think its the only reasonable association. The general guidance with DI is that the types being created from the DI container should not have any tight coupling to the DI container (for example don't have your constructor use IServiceProvider as input argument). This allows using those types without DI because a caller could always manually instantiate the same arguments. This perspective suggests that any of the types that show up in the constructor parameters, such as IMeterFactory, should not be treated as tightly coupled to DI. If the type isn't coupled to DI the fact that it would land in Microsoft.Extensions when all the other Meter types are in System.Diagnostics didn't feel so intuitive any more.

Ultimately I assume its rare for .NET developers to manually create an HttpMessageInvoker so if folks think its not helpful to follow the IMeterFactory pattern on this one I'll be fine with it. If the pattern is also going to show up on HttpClient I might push a little harder due to more exposure, but still I expect the main usage for injecting a Meter will come from HttpClientFactory internally rather than .NET devs manually injecting one.

Reply:
> Ultimately I assume its rare for .NET developers to manually create an HttpMessageInvoker so if folks think its not helpful to follow the IMeterFactory pattern on this one I'll be fine with it. If the pattern is also going to show up on HttpClient

`HttpMessageInvoker` is the base class for `HttpClient` so everything that goes into `HttpMessageInvoker` goes into `HttpClient`. I'm treating them as synonyms in this discussion.

> Define a parameter on your constructor of type IMeterFactory

The pattern with `HttpClient` is to consume configuration/dependencies through properties; the only parameter being passed to the constructor is the `HttpMessageHandler`. It would look very weird if new ctr. overloads would appear taking an `IMeterFactory`, especially if most users don't need it. (In general, `System.Net.Http` wasn't designed to be DI-friendly, so making an occasional intrusive decisions following a different design approach would create too much disharmony IMO.)

>  I expect the main usage for injecting a Meter will come from HttpClientFactory internally rather than .NET devs manually injecting one

In that case, wouldn't it be straightforward to make `HttpClientFactory`'s responsibility to wrap `IMeterFactory` into a `Func<string, Meter>` when initializing `HttpClient`?

> If the type isn't coupled to DI the fact that it would land in Microsoft.Extensions when all the other Meter types are in System.Diagnostics didn't feel so intuitive any more.

Even if it's not tightly coupled to DI, the question is if there are any other reasons to make it land in `System.Diagnostics` other than making DI use-cases easier? Also: are there any other `System.*` libraries that may end up consuming  `IMeterFactory`?

Reply:
> Ultimately I assume its rare for .NET developers to manually create an HttpMessageInvoker so if folks think its not helpful to follow the IMeterFactory pattern on this one I'll be fine with it. If the pattern is also going to show up on HttpClient I might push a little harder due to more exposure, but still I expect the main usage for injecting a Meter will come from HttpClientFactory internally rather than .NET devs manually injecting one.

As @antonfirsov mentioned, `HttpMessageInvoker` = `HttpClient`. The client inherits from the invoker. The design needs to work for both.

Manually creating `HttpClient` is very common. I expect customizing the meter on a client will be much rarer. The primary case for using it will be `Microsoft.Extensions.Http`.

The way I imagined this hanging together is `Microsoft.Extensions.Http` has a dependency on `Microsoft.Extensions.Metrics`. `IMeterFactory` is injected into `HttpClientFactory` and sets a meter on the created HttpClient from the factory.

Reply:
> HttpMessageInvoker is the base class for HttpClient

Ah! I didn't catch that relationship. Thanks for the important info.

> The pattern with HttpClient is to consume configuration/dependencies through properties

OK. If HttpClient already has that pattern then I guess we have clashing patterns, but it probably makes sense for HttpClient's pattern to take priority.

> In that case, wouldn't it be straightforward to make HttpClientFactory's responsibility to wrap IMeterFactory into a Func<string, Meter> when initializing HttpClient

I agree it doesn't take a lot of code to do that wrapping, but assume it was already the plan that IMeterFactory was in System.Diagnostics.DiagnosticSource.dll - would there be any reason to wrap it in the first place?

> are there any other System.* libraries that may end up consuming IMeterFactory?

I don't have any concrete plan to point to, but other forms of IO like files or pipes seems like potential candidates for metrics in the System.* namespaces. Its also possible we'd want System.Runtime metrics reported via a factory-produced Meter. 

> the question is if there are any other reasons to make it land in System.Diagnostics other than making DI use-cases easier

I don't think we need more reasons to meet the bar, but these are the pros/cons as I see it:

- [Pro]: More libraries will have IMeterFactory accessible to them if they aren't able to take a dependency on Microsoft.Extensions. Even in libraries where there isn't a strict rule against such dependencies, library authors are still likely to prefer fewer dependencies.
- [Pro]: Assuming we also have IMeterFactory in the System.Diagnostics.Metrics namespace then consumers only need one namespace in their code rather than two.
- [Con?]: There is the idea that any type used to make the DI scenario easier (IMeterFactory) belongs in M.E.something.dll. However aside from the idea feeling natural, I haven't seen any concrete benefit to organizing it that way in this case.

When I look at the balance scales I see it as small upside and effectively no downside.